Consistent trigger names

rhcarvalho opened the pull request:
Fixes #2594.

rhcarvalho commented:
@bparees could you give some early feedback on this? I want to validate my approach before doing the same for GitHub and Generic.

Doubts:
* Should we change `pkg/build/api/types.go`? That will require adding conversion code to v1beta1 (If yes, I know how to do it, but didn't push it here yet).
* Is it fine to create the constant `ImageChangeBuildTriggerTypeDeprecated`? This allow us to avoid writing magical strings around. This type would be only in v1beta3
* Validation: v1beta1 would accept only "imageChange", v1beta3 accepts both forms, and v1 only "ImageChange". Makes sense?

mfojtik commented:
v1beta1 is now disables
On Jun 3, 2015 7:43 PM, "Rodolfo Carvalho" <notifications@github.com> wrote:

> @bparees <https://github.com/bparees> could you give some early feedback
> on this? I want to validate my approach before doing the same for GitHub
> and Generic.
>
> Doubts:
>
>    - Should we change pkg/build/api/types.go? That will require adding
>    conversion code to v1beta1 (If yes, I know how to do it, but didn't push it
>    here yet).
>    - Is it fine to create the constant
>    ImageChangeBuildTriggerTypeDeprecated? This allow us to avoid writing
>    magical strings around. This type would be only in v1beta3
>    - Validation: v1beta1 would accept only "imageChange", v1beta3 accepts
>    both forms, and v1 only "ImageChange". Makes sense?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/2750#issuecomment-108542896>.
>

bparees commented:
@mfojtik is it? i didn't think quite yet.....

mfojtik commented:
i remember @liggit saying today that we are not exposing v1b1 anymore
On Jun 3, 2015 7:50 PM, "Ben Parees" <notifications@github.com> wrote:

> @mfojtik <https://github.com/mfojtik> is it? i didn't think quite yet.....
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/2750#issuecomment-108545317>.
>

liggitt commented:
it is disabled by default, meaning you can't make requests to /osapi/v1beta1/..., but conversions still exist

rhcarvalho commented on pkg/build/api/v1beta3/conversion.go:
```diff
@@ -336,7 +336,28 @@ func init() {
+			if err := s.DefaultConvert(in, out, conversion.DestFromSource); err != nil {
+				return err
+			}
+			switch in.Type {
```
@bparees if this looks good I'll add other cases in this switch for GitHub and Generic.

mfojtik commented:
@rhcarvalho you also need make this change to v1

rhcarvalho commented:
@mfojtik what change? v1 and internal being both "ImageChange", there's no need for conversion code.

bparees commented:
@rhcarvalho yes this looks like the right approach (ignoring the v1b1 stuff that isn't going to be needed) but i'd like a sanity check from @smarterclayton too, to make sure we're ok not converting v1.  i'm pretty sure we are.

rhcarvalho commented:
@bparees @smarterclayton PTAL

mfojtik commented on pkg/build/api/types.go:
```diff
@@ -340,17 +340,17 @@ type BuildTriggerPolicy struct {
 	// ImageChangeBuildTriggerType represents a trigger that launches builds on
 	// availability of a new version of an image
-	ImageChangeBuildTriggerType BuildTriggerType = "imageChange"
+	ImageChangeBuildTriggerType BuildTriggerType = "ImageChange"
```
you will have to change documentation because of this.

mfojtik commented on pkg/build/api/v1beta1/conversion_test.go:
```diff
@@ -759,3 +759,77 @@ func TestContextDirConversion(t *testing.T) {
+		Old                      older.BuildTriggerPolicy
+		ExpectedBuildTriggerType newer.BuildTriggerType
+	}{
+		"ImageChange": {
```
any reason why not use constants you created?

rhcarvalho replied:
These are just arbitrary test labels used when something goes wrong (see [1](https://github.com/openshift/origin/pull/2750/files#diff-994e87ad6ef744a8680c3b6981c995efR792) & [2](https://github.com/openshift/origin/pull/2750/files#diff-994e87ad6ef744a8680c3b6981c995efR795)).

Same thing as in many other places that use `testCases := map[string]struct`, e.g. [pkg/template/template_test.go](https://github.com/openshift/origin/blob/82458ada79005120d9436cf9dbe1a2f0894101bd/pkg/template/template_test.go#L154-L310).

mfojtik replied:
oh ok

mfojtik commented:
@rhcarvalho can you squash? lgtm (just minor nits)

rhcarvalho commented:
@mfojtik yes, I will squash before we merge :)

Are you fine with renaming the constant from Git**h**ubWebHookBuildTriggerType to Git**H**ubWebHookBuildTriggerType?

And are we keeping or not the conversions for v1beta1?

rhcarvalho commented:
@bparees keep or drop conversion in v1beta1?

bparees commented:
drop the v1b1 conversions but wait for @smarterclayton to delete v1b1 before you merge this.

bparees commented:
@rhcarvalho i'm fine w/ the constant rename and +1 on @mfojtik's doc update comment.

bparees commented:
lgtm. [merge]

rhcarvalho commented:
There was a broken validation test. The test was not very precise and we were validating the same thing twice with different error types and messages. Fixed it.

Also renamed GitHubWebHook and some textual references to GitHub.

I didn't change though the [JSON for key name "github"](https://github.com/openshift/origin/blob/546fda9a6859fe8e9ddb1515ae384f38d78a88ee/pkg/build/api/v1/types.go#L325) (to "gitHub" -- camelCase). @bparees do we change that too for v1 or leave it as is?

rhcarvalho commented:
As for docs, I looked into openshift-docs and found roughly 3 files that mention build trigger types:
* [creating_images/sti_testing.adoc](https://github.com/openshift/openshift-docs/blob/ee6ccf65126f899cd4c894a9246e3d29e3ae9a7f/creating_images/sti_testing.adoc#using-openshift-build-for-automated-testing) -- refers to v1beta1
* [dev_guide/builds.adoc](https://github.com/openshift/openshift-docs/blob/ee6ccf65126f899cd4c894a9246e3d29e3ae9a7f/dev_guide/builds.adoc#build-triggers) -- refers to v1beta1 in some places, and no version in some others
* [architecture/core_objects/builds.adoc](https://github.com/openshift/openshift-docs/blob/ee6ccf65126f899cd4c894a9246e3d29e3ae9a7f/architecture/core_objects/builds.adoc#buildconfig) -- v1beta3

@mfojtik @bparees since the visible changes apply only to v1, how should I approach updating the docs?

bparees commented:
> I didn't change though the JSON for key name "github" (to "gitHub" -- camelCase). @bparees do we change that too for v1 or leave it as is?

@rhcarvalho leave it.
    
> @mfojtik @bparees since the visible changes apply only to v1, how should I approach updating the docs?

pull request against openshift-docs.  but talk to @mfojtik since he's already modifying some of those docs.

[merge]

bparees commented:
@rhcarvalho this still needs a squash.

bparees commented:
[merge]

rhcarvalho commented:
[test]

rhcarvalho commented:
I found [this failure](https://travis-ci.org/openshift/origin/jobs/65859637#L1119-L1124) in the travis log. I ran `hack/test-cmd.sh` locally and get the same error both in this PR's branch and in origin@aa6c8232562 (master).
I don't really know how to tell what went wrong by looking at those logs... :( was it indeed just that test case?

What should I do? @mfojtik @bparees

rhcarvalho commented:
Ops, my mistake, I meant 8cdb79f940

bparees commented:
@rhcarvalho you changed the case of the GitHub grep in test-cmd.sh but i don't think you actually changed the descriptive text in the trigger that it's grepping for.

csrwng commented on pkg/build/webhook/github/github_test.go:
```diff
@@ -234,7 +234,7 @@ func setup(t *testing.T, filename, eventType string) *testContext {
 	req, err := http.NewRequest("POST", "http://origin.com", bytes.NewReader(event))
 	req.Header.Add("Content-Type", "application/json")
-	req.Header.Add("X-Github-Event", eventType)
+	req.Header.Add("X-GitHub-Event", eventType)
```
The github document uses  "X-Github-Event" -> https://developer.github.com/webhooks/

bparees replied:
i'm actually puzzed now, because our code is looking for X-GitHub-Event (even before this PR):
https://github.com/openshift/origin/blob/master/pkg/build/webhook/github/github.go#L101

so 1) i'm not sure how this test ever worked
2) not sure the code in webhook.go was ever correct but it sounds like it should be updated with this PR

maybe headers are case insensitive when doing header.Get()?

liggitt replied:
yes, header-case is canonicalized

smarterclayton replied:
Probably better to revert this part, add a constant and a comment explaining this is EXACTLY what upstream sends.

soltysh replied:
:+1:

soltysh replied:
And change in the code as well to use the one @csrwng mentioned. And please double check what exactly GH sends, just in case.

csrwng commented on pkg/build/webhook/github/github_test.go:
```diff
@@ -90,7 +90,7 @@ func TestWrongContentType(t *testing.T) {
 	req, _ := http.NewRequest("POST", server.URL+"/build100/secret101/github", nil)
 	req.Header.Add("Content-Type", "application/text")
-	req.Header.Add("X-Github-Event", "ping")
+	req.Header.Add("X-GitHub-Event", "ping")
```
X-Github-Event

soltysh replied:
Yeah, you cannot change this, as this is how GH sends those events.

bparees replied:
you can change it, we discussed on IRC, header names aren't case sensitive.  but we should be sending it the way github sends it for consistency.

rhcarvalho replied:
I don't know which way GitHub sends it. They document it as "Github" as @csrwng pointed out. Prior to this PR we had both X-GitHub-Event and X-Github-Event -- let's have only one. Which one?

soltysh replied:
The one from their doc.

bparees replied:
which means putting it back to X-Github-Event

soltysh replied:
I was correct with this, check just in case, because apparently in the docs they have lowercased version but they are sending this: `X-GitHub-Event: push`

rhcarvalho replied:
What?! Both forms are fine, I'm removing changes to X-Github-Event from this PR and we can make its usage consistent across the repo later. Afterall, this PR is about trigger names.

csrwng commented on pkg/build/webhook/github/github_test.go:
```diff
@@ -149,7 +149,7 @@ func TestJsonPushEventError(t *testing.T) {
 	post("X-GitHub-Event", "push", []byte{}, server.URL+"/build100/secret101/github", http.StatusBadRequest, t)
```
X-Github-Event

csrwng commented on test/integration/webhookgithub_test.go:
```diff
@@ -270,7 +270,7 @@ func postFile(client kclient.HTTPClient, event, filename, url string, expStatusC
 	req.Header.Add("Content-Type", "application/json")
 	req.Header.Add("User-Agent", "GitHub-Hookshot/github")
-	req.Header.Add("X-Github-Event", event)
+	req.Header.Add("X-GitHub-Event", event)
```
X-Github-Event

rhcarvalho commented:
Given that HTTP headers are case insensitive and that GitHub calls it "X-Github-Event", what do we want to have in our repo, "X-Github-Event" or "X-GitHub-Event"?

smarterclayton commented:
This is a breaking change for v1beta3 and v1 because of the case change for constants - if you do this, you break all clients out there. v1beta3 and v1 will have to accept lowercase values (the window for change to v1 has closed).

rhcarvalho commented:
@bparees the problem was using raw strings rather than the constants for trigger types https://github.com/openshift/origin/pull/2750/files#diff-34571ad9de2d354ca8f3703bf95710efL132

rhcarvalho commented:
@smarterclayton this PR is changing the cases for v1 and internal format, v1beta3 stays as it was.

Should we accept both "generic"/"Generic", "github"/"GitHub" and "imageChange"/"ImageChange" in v1?

smarterclayton commented:
If this is closed in the next few hours we can land it. @bparees this is your call.

----- Original Message -----
> @smarterclayton this PR is changing the cases for v1 and internal format,
> v1beta3 stays as it was.
> 
> Should we accept both "generic"/"Generic", "github"/"GitHub" and
> "imageChange"/"ImageChange" in v1?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2750#issuecomment-110006654

bparees commented:
@smarterclayton This PR already accepts both formats for v1b3 due to the conversion, unless i'm missing something....

For V1, i'm not averse to adding conversion for V1 also if that's what needs to be done, but i do want the PR to land.

smarterclayton commented:
Just want to be absolutely sure.  Talking about changing web hooks scared me.  Remember, web hooks are in URLs which are all lower case.

bparees commented:
@rhcarvalho make sure this hasn't broken the webconsole since it also displays triggers and might have similar logic to what describer was doing.

rhcarvalho commented:
@bparees I added tests to ensure v1beta3 behaves as you said, and conversion code to make v1 behave similarly.

If you're fine with those changes and tests pass I'll squash commits.

rhcarvalho commented:
@smarterclayton this PR is not changing web hook URLs.

smarterclayton commented:
Approved for merge on ben's recognizance

soltysh commented on pkg/client/testclient/fake_buildconfigs.go:
```diff
@@ -33,8 +33,8 @@ func (c *FakeBuildConfigs) WebHookURL(name string, trigger *buildapi.BuildTrigge
-	case trigger.GithubWebHook != nil:
-		return url.Parse(fmt.Sprintf("http://localhost/buildConfigHooks/%s/%s/github", name, trigger.GithubWebHook.Secret))
+	case trigger.GitHubWebHook != nil:
+		return url.Parse(fmt.Sprintf("http://localhost/buildConfigHooks/%s/%s/github", name, trigger.GitHubWebHook.Secret))
```
It would be good to change those fake client to also follow the current URL pattern.

rhcarvalho replied:
@soltysh not in this PR

soltysh replied:
I'm fine with a follow-up.

bparees commented on pkg/build/api/v1beta3/conversion_test.go:
```diff
@@ -1,11 +1,91 @@
 
 	knewer "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
-	kolder "github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1beta3"
+	/*kolder "github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1beta3"*/
```
delete this.

bparees commented:
@rhcarvalho need you to put the X-Github-Event headers back to that form (lowercase h), delete the commented out import, and squash and we can merge.

rhcarvalho commented:
@bparees removed code related to X-Github-Event (we can decide about that later), deleted the commented out import and squashed.

bparees commented:
@rhcarvalho thanks!  @smarterclayton ok i'm signing off on this.

bparees commented:
[merge]
